### PR TITLE
Make `first` and `last` always return 1 element, let `take` return from end of list

### LINKED ISSUE #7128: Proposal: make `first` and `last` always return 1 element, let `take` return from end of list
### Question

Following some discussion on Discord and in [this issue](https://github.com/nushell/nushell/pull/6893), I would like to propose the following changes:

1. Remove the rows parameter from `first` and `last` so they _always_ return a single item
    1. Makes the commands' return type consistent 
    1. Brings Nu's behaviour in line with many other languages
2. Add negative number support to `take` so `take -3` returns the last 3 elements.
    1. Inspired by Python's negative indexing (and has been suggested a few times by Python users)
    1. Lets us support existing `last n` functionality without adding a new command

# Examples

```bash
[0 1 2] | first 2 # will no longer work
[0 1 2] | take 2  # will need to do this instead

[0 1 2] | last 2  # will no longer work
[0 1 2] | take -2 # will need to do this instead
```

# Language Comparison
| **Functionality**             | **First element** | **n-th element** | **Last element** | **First n elements** | **Last n elements**               |
| ----------------------------- | ----------------- | ---------------- | ---------------- | -------------------- | ------------------------------    |
| **Return Type**               | `T`               | `T`              | `T`              | `List<T>`            | `List<T>`                         |
| **Python** `(more_)itertools` | `first `          | `nth`            | `last`           | `take(n)`            | `arr[-n:]`, `tail(n)`             |
| **Javascript** (`lodash`)     |  `head()`         |  `nth`           | `last()`         |  `take(n)`           | `takeRight(n)`                    |
| **.NET / C# / LINQ**          |  `First()`        |  `ElementAt()`   | `Last()`         |   `Take(n)`          | `TakeLast(n)`, `array[^n..]`      |
| **Rust**                      | `first()`         | `nth()`          |  `last()`        |    `take(n)`         | `v[v.len()-3..]`?                 |
| **Scala**                     | `head`            | `drop(n-1).next` | `last`           | `take(n)`            | `.takeRight(n)`                   |
| **Nushell (current)**         | `first`   | `get n`          | `last`           |  **`first n`, `take n`**  |     **`last n`**                      |
| **Nushell (proposed)**        | `first`           | `get n`          | `last`           | **`take n`**             |     **`take -n`**                     |

## Next Steps

If this isn't too controversial I can implement it soon.

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by rgwood**
Closes #7128. This PR does 2 things:

1. Removes the rows parameter from `first` and `last` so they _always_ return a single item
    1. This makes the commands' return type consistent and brings Nu's behaviour in line with many other languages
2. Adds negative number support to `take` so `take -3` returns the last 3 elements.
    1. Inspired by Python's negative indexing (and has been suggested a few times by Python users)
    1. Lets us support existing `last n` functionality without adding a new command

# Examples

```bash
[0 1 2] | first 2 # will no longer work
[0 1 2] | take 2  # will need to do this instead

[0 1 2] | last 2  # will no longer work
[0 1 2] | take -2 # will need to do this instead
```

# Language Comparison
| **Functionality**             | **First element** | **n-th element** | **Last element** | **First n elements** | **Last n elements**               |
| ----------------------------- | ----------------- | ---------------- | ---------------- | -------------------- | ------------------------------    |
| **Return Type**               | `T`               | `T`              | `T`              | `List<T>`            | `List<T>`                         |
| **Python** `(more_)itertools` | `first `          | `nth`            | `last`           | `take(n)`            | `arr[-n:]`, `tail(n)`             |
| **Javascript** (`lodash`)     |  `head()`         |  `nth`           | `last()`         |  `take(n)`           | `takeRight(n)`                    |
| **.NET / C# / LINQ**          |  `First()`        |  `ElementAt()`   | `Last()`         |   `Take(n)`          | `TakeLast(n)`, `array[^n..]`      |
| **Rust**                      | `first()`         | `nth()`          |  `last()`        |    `take(n)`         | `v[v.len()-3..]`?                 |
| **Scala**                     | `head`            | `drop(n-1).next` | `last`           | `take(n)`            | `.takeRight(n)`                   |
| **Nushell (current)**         | `first`   | `get n`          | `last`           |  **`first n`, `take n`**  |     **`last n`**                      |
| **Nushell (proposed)**        | `first`           | `get n`          | `last`           | **`take n`**             |     **`take -n`**                     |

### CODE REVIEW

**Comment by rgwood** [comment]
I forgot about dataframes, sigh. I'm not sure I know enough about dataframes to fix those tests but I will take a look. It seems like the dataframe variant of `last` defaults to 5 rows, odd.

**Comment by rgwood** [comment]
Well, it looks like the dataframe version of `take` works completely differently from the regular version of `take`. This PR is on hold until I can figure out what to do about that. https://www.nushell.sh/book/commands/take.html#frontmatter-title-for-dataframe

**Comment by kubouch** [comment]
Honestly, I think `first 2` and `last 2` are much more descriptive than `take 2` and `take -2`. I'd be OK with having `take` instead of `first` but `take -2` feels too arcane and I think `last` is superior here (see the last column in the table, it's noticeably more complicated than the rest of the table in other languages). And if we have `last`, it is natural to have `first`. Another option would be to have `take` and `tail` but I'm just thinking... there is nothing wrong with `first` and `last` taking an argument?

One problem is having both `take` and `first` working the same way. Since dataframe's `take` works kind of like the current `get`, we could rename it to `get` and also move all the `take while` and `take until` under `get` . This would get rid of the `take` command completely, avoiding the `first`/`take` discrepancy and having one less command to worry about.

**Comment by fdncred** [comment]
@rgwood after looking at the documentation that you linked to, it seems like the dataframe take is different enough to allow it to be named something differently such as `take indices` or `take at-index` or `take index` or even rename entirely to `get` or `get blah` like Kubouch suggests.

**Comment by dandavison** [comment]
@kubouch I agree that `take -2` is a bit arcane. And I also agree that `first n` and `last n` read nicely, although that is tempered by the fact that `take` is very widely-used in other languages.

However: Sorry if I'm misunderstanding. Maybe you're suggesting eliminating the argument-less forms `first` and `last`? But if not:

> there is nothing wrong with first and last taking an argument?

So in that case, `first` would be overloaded with the following signatures:

`iterable<T> | first => T`
`iterable<T> | first n => iterable<T>`

and the same for `last`.

@rgwood and I had been working on the understanding that Nushell did not want this form of overloading: i.e. that `(input_type, command_name)` should map uniquely to `output_type`; the  presence/absence of an argument should not alter the output type. Recently there were some design notes written on Types and Overloading. Doesn't what your suggesting contradict the [Overloading](https://hackmd.io/izohHejpS-iNZ6iXz2Clig#Overloading) section?

> The signature and output type must be the same for all overloads. While technically, we could allow overloads also on the signature and output types, it would complicate things.

I don't think I'm aware of other languages taking an approach similar to this (`first => T`, `first n => iterable<T>`) in their standard library of utilities for working with iterables.

**Comment by kubouch** [comment]
I'm suggesting that both `first` and `first 2` keep working and you're right, it messes up the output type and you're right that's a problem. `first` outputs a T but `first 1` an iterable of T, that's not very good. Similar for `last`.

Then, I take my suggestion back but I would still replace `take -2` with `tail 2` or something.

Same philosophy could be applied to the dataframe commands: If it behaves more like `get`, let's call it `get`, or change the behavior to match `take`.

**Comment by webbedspace** [comment]
Quick list of my opinions:
* I don't like multiple return types on commands, and think this is an easy nail to hammer.
* I like negative indexes as a way to access-from-end. The only "problem" is that more Nushell commands don't support it yet. 🦚
* I don't really get why we need to keep commands that *only* access the first or last element, when `get 0` already exists and uses almost the same number of characters. However, I can tolerate it as a pure syntactic sugar thing (or as a pair with `last`).
* The dataframe commands are kind of a huge naming mess. Have you seen `drop`? Compare and contrast that with the real `drop`.

**Comment by rgwood** [comment]
> I would still replace `take -2` with `tail 2` or something.

I'm open to that; personally I'm indifferent between `take -2` and `tail 2`. The negative indexing feels a little arcane, but I agree with @sholderbach's comment "the old pair `head` and `tail` carries too many assumptions and shadowing only one feels icky" from Discord.

> However, I can tolerate it as a pure syntactic sugar thing (or as a pair with `last`).

Yeah, I think those are the main motivations for `first`. I find the syntactic sugar useful, and it looks like most other languages have gone the same way.

**Comment by rgwood** [comment]
Where this is at right now (IMO):

- we have agreement on the `first`+`last` changes
- the `take -2` change is a little more controversial, we're blocked on that

**Comment by fdncred** [comment]
my vote is to land it with the `take -2` functionality. However, at some point, we should standardize our "range" functionalities in commands. I realize that `take` doesn't exactly take a range, but maybe at some point it should.

It seems like `str substring` uses 3 methods.
```
Examples:
  Get a substring "nushell" from the text "good nushell" using a range
  >  'good nushell' | str substring 5..12

  Alternately, you can pass in a list
  >  'good nushell' | str substring [5 12]

  Or a simple comma-separated string
  >  'good nushell' | str substring '5,12'

  Drop the last `n` characters from the string
  >  'good nushell' | str substring ',-5'

  Get the remaining characters from a starting index
  >  'good nushell' | str substring '5,'

  Get the characters from the beginning until ending index
  >  'good nushell' | str substring ',7'

```

File: crates/nu-command/src/filters/take/take_.rs
Comment (sholderbach) [inline crates/nu-command/src/filters/take/take_.rs:88]: This uses excessive memory and should be replaced by a `VecDeque` ring buffer as suggested in #7178

File: crates/nu-command/src/filters/take/take_.rs
Comment (sholderbach) [inline crates/nu-command/src/filters/take/take_.rs:119]: Here using the ring buffer solution would obviate the need for the superfluous collect as well.

File: crates/nu-command/src/filters/take/take_.rs
Comment (sholderbach) [inline crates/nu-command/src/filters/take/take_.rs:102]: Same as above (ref #7178)

**Review by sholderbach — CHANGES_REQUESTED** [review]
Prompted by #7178 that would be made obsolete by this change, there are unnecessary allocations of the whole pipeline in the current code.

File: crates/nu-command/src/filters/take/take_.rs
Comment (rgwood) [reply crates/nu-command/src/filters/take/take_.rs:88]: Good catch! Will do.

**Comment by rgwood** [comment]
@sholderbach Fixed, thanks. Sorry for the force push - was having some trouble with merge conflicts.

**Review by sholderbach — APPROVED** [review]
Thanks! I missed that.

**Comment by rgwood** [comment]
We're approaching the 0.72 release and it's already a very big release. Let's wait until after the release to merge this (assuming we have consent for the `take` change)

**Comment by rgwood** [comment]
Discussed this in the weekly meeting and I think we're not going to land this as-is.

We're not 100% sure about the `take` change, and there are concerns about wanting a consistent approach to get the last `n` items across multiple commands.

More work is needed in this area, but we probably don't want to go with this exact approach.

**Comment by rgwood** [comment]
Some scattered thoughts about where to go with this next:

### Ranges

Apparently `range` can already get the last n rows! For example, `[0,1,2,3,4,5] | range (-2..)` returns `[4,5]`. tbh I'm not a huge fan of that exact syntax. `(-2..)` on is a numeric range from -2 to infinity; IMO it's not obvious that that range corresponds to the last 2 rows. But I like the general idea of using ranges.

I think C#'s `^` [index-from-end operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#systemindex) is pretty good, for example `arr[^2..]` returns the last 2 elements in `arr`. Maybe we could adopt that in Nu: `take ^2..`? `range ^2..`?

A related issue from last year: https://github.com/nushell/nushell/issues/3378

### How strict do we need to about type consistency?

Jakub raised a good question in today's meeting: how strict do we need to be about always returning the same data type from a command? Maybe it's OK if `first` sometimes returns `T` and sometimes `list<T>`?

I think we generally want to be consistent with return types but I'm cautiously open to having a few exceptions like `first` and `last`. Ideally we figure out a great type-consistent design, but we haven't yet landed on one that everyone loves.

**Pull request closed without merging**
